README.md
<p align="center">
<img src="./docs/_images/PVJ_2019.svg" alt="PocketVJ-Logo" width="100%">
</p>


# PocketVJ Exhibition CP 4.x

[![commits](https://img.shields.io/github/last-commit/magdesign/PocketVJ-CP-exh/master.svg?style=flat-square)](https://github.com/magdesign/PocketVJ-CP-exh/commits/master)
[![issues](https://img.shields.io/github/issues-raw/magdesign/PocketVJ-CP-exh.svg?style=flat-square)](https://github.com/magdesign/PocketVJ-CP-exh/issues)
[![pulls](https://img.shields.io/github/issues-pr-raw/magdesign/PocketVJ-CP-exh.svg?style=flat-square)](https://github.com/magdesign/PocketVJ-CP-exh/pulls)


 (not compatible with any other PocketVJ Version!!)

## 📖️ Instructions

See [here](https://video.pocketvj.com/AVideo/video/7/pocketvj-exhibition-update) on how to update your PocketVJ Exhibition.

## 📝 Release Notes

See [**Release notes**](RELEASE_NOTES.md)

## 🚀 Roadmap

See [**Roadmap**](ROADMAP.md)

## ⁉️ Tutorials
Watch [**Videos**](https://www.pocketvj.com/pocketvj-exhibition#Tutorials_Exhibition)


or


Read [**Manual**](https://www.pocketvj.com/manual/index.html)


## 📝 Notes

Exhibition models from the very first generation (4.0.0) need to be updated [on the alternative way](https://video.pocketvj.com/AVideo/video/6/pocketvj-exhibition-alternative-update).
and then go to "system settings" and click "update all", this is important!





RELEASE_NOTES.md
# PocketVJ Exhibition / Control Panel 4.x.x

## added:
- rc.local.randomaudio (for autostart into randomaudio gpio listener)
- randomaudiotrigger.py (listens to gpio3 input to random trigger audios)
- startrandomaudiotrigger (starts the random audio and defines playtime until listen again)

## CP 4.3.4 19.07.2022
- added randomaudio player to buttons
- added customfunction1+2 to buttons
- updated stopall to kill randomaudio player
- customsfunction1 launches the mapper and kills it after n amount of time

## CP 4.3.3b 03.06.2022
 
 - removed ugly rental reset sign, since it was more distracting than helpful to users<br />

## CP 4.3.3a 11.03.2022

- better mouse detection for mapping mode<br />
- added mapperpreset 7-10<br />
- added OSC control to load mapperpreset 1-10<br />
- added OSC control to save mapperpreset 1-10<br />

## CP 4.3.3 22.02.2022

- powersupply check did not show every state, fixed now!<br />
- removed mapping converter<br />
- added save painter as overlay<br />
- fixed issue on "Test Connection" under "Network"<br />
- better IP adress on screen display when using "Test Connection"<br />

## CP 4.3.2a 14.02.2022

- added middle mouse scroll to painter to adjust thickness (this is soo cool!)<br />

## CP 4.3.2 23.01.2022

- added simple painter to easy paint masks<br />

## CP 4.3.1a 15.01.2022

- moved rental reset to a script, now getting propper feedback<br />
- fixed typos in hyperion install<br />
- added propper service for hyperion daemon<br />
- force audio to hdmi when using hyperion<br />
- set audio to hdmi/jack now also switches on the mapper<br />
- updated manual<br />

## CP 4.3.1 14.01.2022

- added hyperion (and all debs) for led strip ambient lights<br />
- fixed an issue  in enable lirc<br />

## CP 4.3.0b 11.01.2022

- fixed updateall so you dont have to do it 2x, first copying stuff, then executing<br />


## CP 4.3.0a 06.01.2022

- resolution change now works without reboot<p/>
- added stopall command to reboot and poweroff to prevent sd card corruption<p/>
- fix little error in updateall script to disable logging<p/>

## CP 4.3.0 04.01.2022

- changed layout in DMX tab, so recording does not show up in the beginning (client feedback)<br />
- OLA now disables and stops itself before executing updateola<p/>
- "Update Everything" is now in a seperate script in /var/www/sync/updateall<p/>
- "Update All" & "Factory Reset" now disables logging <p/>
- Added "rescue" script to all autostarts, to make sure webserver can always run <p/>


## CP 4.2.9b 28.12.2021

- added "overlay" to "autostart" => "master" and to "player sync" => play all <br />
- there might be some edge cases when overlay is on and you enable it again... <br />

## CP 4.2.9 30.11.2021

- made Check Services (ping) much faster with setting a timout <br />
- added startaudio01-03 once scripts <br />
- added startaudio once to buttons <br />
- updated the audio output selection to include new scripts  <br />
- remove autostart beacon&buttons, since it is not used <br />
- added startmasterone to scheduler <br />
- addED scheduler status to "Check Services"  <br />

## Image 4.0.10 04.November 2021 <br />

- updated the GPIO library to gpiozero<br />

## CP 4.2.8 04.11.2021

after installing the update, click "Update Everything" again!<br />

- added gpiozero library <br />
- changed the button trigger script, finally with propper debouncing when pushed very very very fast...<br />

## CP 4.2.7 27.10.2021

- changed button trigger for dmx shows to only play once <br />
- added scripts: stardmxplaybackaudio02once, stardmxplaybackaudio03once <br />

## CP 4.2.6 26.10.2021

- Button 01 and Button 03 where wrong numbered, flipped them <br />
- Increased debounce time of buttons to 4000ms <br />
- After changing button function when already using them, they did not get active, now it does <br />
- "Autostart Custom" now starts dmxshow/audio01_* once and goes to button listener <br />
- Added script stardmxplaybackaudio01once <br />

## CP 4.2.5 26.10.2021

- fixed audio was not stopping when switching dmx show <br />
- added autostart buttons together with beacons <br />
- added play dmxshow 01 + 02 only once <br />

## CP 4.2.4a 25.10.2021
!! needs 2x "Upate Everything" !!

- added audio boost to Set Audio to 100% on all channels <br />

## CP 4.2.4 24.10.2021
!! needs 2x "Upate Everything" !!

- massive increase of DMX audio quality with changing the audio player<br />
- added new audioplayer kill to stopall script<br />
- changes in play audio script (stopall kills now everything)<br />
- factory reset and update all will set audio to jack output, since setting to both decreases audio quality a lot  <br />
- testtone now works according to the audio output selection, including usb audio <br />
- added dmx/audio to buttons  <br />

## CP 4.2.3 16.October 2021

- fixed dmx blackout when starting show 05-10  <br />
- fixed autostart to dmx show, added blackout and longer sleep<br />
- tweaked usb mount <br />
- added autostart master usb (hope this works also with sticks from OSX & win)<br />
- fixed execution rights on servicecheck<br />

## CP 4.2.2 24.September 2021

- added SLIDESHOW after to play slideshow after video played once<br />
- fixed error in pyGame installer<br />
- added python-zmq to be installed when installing pygame<br />
- added first sketch of grandPA.py, there is a hell lot to learn, any help welcome<br />
- added docs to system, so they can be read offline (prepare to bring everything offline available)<br />

## CP 4.2.1a  30.August 2021  <br />

- fixed stopall/setdmxblackout to not reset our controluniverse 15<br />

## CP 4.2.1  26.August 2021  <br />

- added webcam with HSV effect<br />
- added very experimental gender detection (might overheat your unit!)<br />

## CP 4.2.0a  25.August 2021  <br />

- fixed selector for beacon functions <br />
- added scale slider to video  <br />

## CP 4.1.10  23.August 2021  <br />

sync/gender.py
# origin: https://gist.github.com/magdesign/06434a4412b6f378199f4e40731fc238
#made for pocketvj exhibition
# still a few issues: 1. opens too many windows, 2.is very laggy
#run it like this:

#python3 /var/www/sync/gender.py --image_source=webcam --pygame_display=yes --opencv_display=yes --delay=1 --haar_path=/home/pvj/opencv/opencv-3.4.1/data/haarcascades/haarcascade_frontalface_alt.xml
#python3 /var/www/sync/gender.py --image_source=webcam --pygame_display=none --opencv_display=yes --delay=1 --haar_path=/home/pvj/opencv/opencv-3.4.1/data/haarcascades/haarcascade_frontalface_alt.xml



import pygame
import pygame.camera
from pygame.locals import *

import time
import imutils
import cv2
import numpy as np
import argparse

# http://www.pygame.org/wiki/HeadlessNoWindowsNeeded
# https://stackoverflow.com/questions/10466590/hiding-pygame-display  (Chris Jeong's answer)
import os
#os.environ["SDL_VIDEODRIVER"] = "dummy"

# allow the camera to warmup
time.sleep(0.1)

DEVICE   = '/dev/video0'
SIZE     = (640, 480)
FILENAME = 'capture.png'


MODEL_MEAN_VALUES = (78.4263377603, 87.7689143744, 114.895847746)
age_list          = ['(0, 2)','(4, 6)','(8, 12)','(15, 20)','(25, 32)','(38, 43)','(48, 53)','(60, 100)']
gender_list       = ['Male', 'Female']


def getargs():
    '''
    Get arguments. There is one argument: --imagesource which can be either 'nachiket' or 'webcam'
    '''
    parser = argparse.ArgumentParser(description='Argument parser for face recognition')
    parser.add_argument('--image_source',help='defines source of images for face recognition',
                        type=str,required=True,default='webcam',choices=['webcam','nachiket']
                       )
    parser.add_argument('--haar_path',help='defines absolute path of haarcascade_frontalface_alt.xml including filename',
                        type=str,required=False,default=None
                       )
    parser.add_argument('--pygame_display',help='turn on or off the pygame display',type=str,required=False,default='yes',
                        choices=['yes','no']
                       )

    parser.add_argument('--opencv_display',help='turn on or off the opencv display',type=str,required=False,default='yes',
                        choices=['yes','no']
                       )

    parser.add_argument('--delay',help='specify time delay between refreshing frames (millisecs)',required=False,default=1,type=int)
                        

    
    args = parser.parse_args()

    # process args
    #if (args.haar_path == 'None'):
    #    args.haar_path = None

    # turn off display for pygame 
    if (args.pygame_display == 'no'):  os.environ["SDL_VIDEODRIVER"] = "dummy"
    if (args.haar_path == 'None'):     args.haar_path = None

    assert (args.delay>0),'--delay should be integer > 0, you specified {}'.format(args.delay)
        
    return args

 
def initialize_caffe_model():
    # models come from 
    # https://itywik.org/2018/03/26/age-and-gender-detection-with-opencv-on-the-raspberry-pi/
    # http://itywik.org/download/128/
    print('Loading models...')
    age_net = cv2.dnn.readNetFromCaffe(
                        "/home/pvj/opencv/age_gender_models/deploy_age.prototxt", 
                        "/home/pvj/opencv/age_gender_models/age_net.caffemodel")
    gender_net = cv2.dnn.readNetFromCaffe(
                        "/home/pvj/opencv/age_gender_models/deploy_gender.prototxt", 
                        "/home/pvj/opencv/age_gender_models/gender_net.caffemodel")

    print('Models loaded...')
 
    return (age_net, gender_net)

def image_nachiket():
    '''
    A generator function which reads nachiket.jpg and yields it forever
    This generator is used for testing if the cv face-recognition actually works
    image is in bgr format because cv2 reads it in bgr format
    https://stackoverflow.com/questions/54959387/rgb-image-display-in-matplotlib-plt-imshow-returns-a-blue-image
    '''
    image = cv2.imread('nachiket.jpg',1)
    image = np.array(image)

    while True:
        yield image


def image_generator(device=DEVICE,size=SIZE,filename=FILENAME,pygame_display='yes'):
    '''
    A generator function which scans device and yields images
    '''
    
    # initialize pygame
    pygame.init()
    pygame.camera.init()
    display = pygame.display.set_mode(SIZE,0)
    pygame.mouse.set_visible(0)  # hide mouse
    camera = pygame.camera.Camera(DEVICE,SIZE)
    camera.start()
    screen = pygame.surface.Surface(SIZE,0,display)
    
    capture = True
    while capture:
        # gets RGB image - need to convert RGB to BGR for cv2
        # https://www.pygame.org/docs/ref/camera.html#pygame.camera.Camera.get_image        
        screen = camera.get_image(screen)
        if pygame_display == 'yes':
            # blit means to draw objects in the screen buffer to display        
            display.blit(screen,(0,0))
            # update the display 
            pygame.display.flip()
            
        # https://stackoverflow.com/questions/34673424/how-to-get-numpy-array-of-rgb-colors-from-pygame-surface
        s3d=pygame.surfarray.array3d(screen)
        # swap the first two axes
        s3d=np.swapaxes(s3d,0,1)  

        # IMPORTANT
        # convert rgb to bgr DO NOT use: s3d=s3d[:,:,::-1]
        # https://github.com/opencv/opencv/issues/18120#issuecomment-688190878
        # cv2.COLOR_RGB2BGR and COLOR_BGR2RGB both evaluate to 4
        s3d = cv2.cvtColor(s3d,cv2.COLOR_RGB2BGR)
        
        yield s3d

def capture_loop(age_net,gender_net,imgen,haar_path=None,opencv_display='no',delay=1):
    '''
    This is the capture loop.
    age_net, gender_net are dnns (created by initialize_caffe_model)
    imgen is a generator which produces images (bgr) to feed to age_net and gender_net
    '''
    print('Starting capture loop...')
    font = cv2.FONT_HERSHEY_SIMPLEX
    # https://stackoverflow.com/questions/30508922/error-215-empty-in-function-detectmultiscale        
    # face_cascade = cv2.CascadeClassifier('/home/pi/venv/magopencv/lib/python3.7/site-packages/cv2/data/haarcascade_frontalface_alt.xml')
    # face_cascade = cv2.CascadeClassifier('haarcascade_frontalface_alt.xml')
    # print('haar_path=',type(haar_path))
    if haar_path is None:
        face_cascade = cv2.CascadeClassifier(cv2.data.haarcascades+'haarcascade_frontalface_alt.xml')
    else:
        face_cascade = cv2.CascadeClassifier(haar_path)
    
    for image in imgen:
        np.save('input_from_camera',image)
        gray  = cv2.cvtColor(image,cv2.COLOR_BGR2GRAY)
        faces = face_cascade.detectMultiScale(gray, 1.1, 5)
        print("Found ",len(faces)," face(s)")
        
        for (x,y,w,h) in faces:
            # https://github.com/opencv/opencv/issues/18120
            cv2.rectangle(image,(x,y),(x+w,y+h),(255,255,0),2)
            np.save('face_detected',image)
            face_img = image[y:y+h, x:x+w].copy()
            blob = cv2.dnn.blobFromImage(face_img,1,(227, 227), MODEL_MEAN_VALUES, swapRB=False)
            # Predict gender
            gender_net.setInput(blob)
            gender_preds = gender_net.forward()
            gender = gender_list[gender_preds[0].argmax()]
            # Predict age
            age_net.setInput(blob)
            age_preds = age_net.forward()
            age = age_list[age_preds[0].argmax()]
            overlay_text = "%s, %s" % (gender, age)
            cv2.putText(image, overlay_text ,(x,y), font, 0.6,(255,255,255),2,cv2.LINE_AA)
            print(gender, age)

        if (opencv_display == 'yes'): cv2.imshow("Image",image)
        key = cv2.waitKey(delay) 
        # if (opencv_display == 'yes'): cv2.destroyAllWindows()

if __name__ == '__main__':
    
    args    = getargs()
    gendict = {
               'webcam':image_generator(device=DEVICE,size=SIZE,filename=FILENAME,
                                        pygame_display=args.pygame_display
                                       ),
               'nachiket':image_nachiket()
              }
    imgen   = gendict[args.image_source]
    
    age_net,gender_net = initialize_caffe_model()     # initialize models, pygame
    #imgen  = image_generator(device=DEVICE,size=SIZE,filename=FILENAME) # streams usb camera
    #imnach = image_nachiket()  # this produces only a stream of nachiket.jpg
    capture_loop(age_net=age_net,gender_net=gender_net,
                 imgen=imgen,haar_path=args.haar_path,
                 opencv_display=args.opencv_display,
                 delay=args.delay
                )
    
        
    


